From 1c23b23c6b2ada081e630893534f75e2d8324f75 Mon Sep 17 00:00:00 2001
From: Yong Zhang <yong.zhang@windriver.com>
Date: Sun, 18 Apr 2010 21:34:44 -0700
Subject: [PATCH 07/11] net: tracking high water mark for per-device transmit
 queue

Enabling this option will enable tracking the high water mark for
the per-device transmit queues.  The high water mark of the queue
will be exported via /proc/net/dev if procfs is enabled.

Signed-off-by: Yongli he   <yongli.he@windriver.com>
Integrated-by: Yong Zhang <yong.zhang@windriver.com>
Signed-off-by: Joe MacDonald <joe.macdonald@windriver.com>
---
 include/net/sch_generic.h |   27 +++++++++++++++++++++++++++
 net/Kconfig               |   10 ++++++++++
 net/core/dev.c            |   12 ++++++++----
 net/sched/sch_generic.c   |    4 ++--
 4 files changed, 47 insertions(+), 6 deletions(-)

diff --git a/include/net/sch_generic.h b/include/net/sch_generic.h
index 9d7d54a..78ed2b5 100644
--- a/include/net/sch_generic.h
+++ b/include/net/sch_generic.h
@@ -14,6 +14,32 @@ struct qdisc_walker;
 struct tcf_walker;
 struct module;
 
+#ifdef CONFIG_QDISC_STATS
+#define HIGHQLEN_DECL u32 highqlen;
+#define check_highqlen(qdisc)				\
+do {							\
+	unsigned int qlen = qdisc->q.qlen;		\
+	if (qlen > qdisc->highqlen)			\
+		qdisc->highqlen = qlen;			\
+} while (0)
+
+#define dev_seq_printf_qdisc_stuff(seq, dev)		\
+do {							\
+	if (netdev_get_tx_queue(dev, 0)->qdisc)		\
+		seq_printf(seq, "%10u",			\
+		netdev_get_tx_queue(dev, 0)->qdisc->highqlen); \
+} while (0)
+#define dev_seq_show_qdisc_stuff(seq)			\
+	do {seq_puts(seq, " qdhighmark"); } while (0)
+#else
+#define HIGHQLEN_DECL
+#define check_highqlen(qdisc)
+#define dev_seq_printf_qdisc_stuff(seq, dev)
+#define dev_seq_show_qdisc_stuff(seq)
+#endif
+#define qdisc_incr_qlen(qdisc)				\
+do {qdisc->q.qlen++; check_highqlen(qdisc); } while (0)
+
 struct qdisc_rate_table {
 	struct tc_ratespec rate;
 	u32		data[256];
@@ -68,6 +94,7 @@ struct Qdisc {
 	 * and it will live until better solution will be invented.
 	 */
 	struct Qdisc		*__parent;
+	HIGHQLEN_DECL
 	struct netdev_queue	*dev_queue;
 	struct Qdisc		*next_sched;
 
diff --git a/net/Kconfig b/net/Kconfig
index f302f49..9bbd11b 100644
--- a/net/Kconfig
+++ b/net/Kconfig
@@ -50,6 +50,16 @@ source "net/unix/Kconfig"
 source "net/xfrm/Kconfig"
 source "net/iucv/Kconfig"
 
+config QDISC_STATS
+	default n
+	bool "Qdisc statistics reporting"
+	help
+	  Enabling this option will enable tracking the high water mark for
+	  the per-device transmit queues.  The high water mark of the queue
+	  will be exported via /proc/net/dev if procfs is enabled.
+
+	  Say Y if you want this additional tracking.
+
 config DGRAM_SOCKSTATS
 	bool "Datagram socket statistics"
 	default n
diff --git a/net/core/dev.c b/net/core/dev.c
index 9ed8b4f..8c4e9dd 100644
--- a/net/core/dev.c
+++ b/net/core/dev.c
@@ -4105,7 +4105,7 @@ static void dev_seq_printf_stats(struct seq_file *seq, struct net_device *dev)
 	const struct rtnl_link_stats64 *stats = dev_get_stats(dev, &temp);
 
 	seq_printf(seq, "%6s: %7llu %7llu %4llu %4llu %4llu %5llu %10llu %9llu "
-		   "%8llu %7llu %4llu %4llu %4llu %5llu %7llu %10llu\n",
+		   "%8llu %7llu %4llu %4llu %4llu %5llu %7llu %10llu",
 		   dev->name, stats->rx_bytes, stats->rx_packets,
 		   stats->rx_errors,
 		   stats->rx_dropped + stats->rx_missed_errors,
@@ -4121,6 +4121,8 @@ static void dev_seq_printf_stats(struct seq_file *seq, struct net_device *dev)
 		    stats->tx_window_errors +
 		    stats->tx_heartbeat_errors,
 		   stats->tx_compressed);
+	dev_seq_printf_qdisc_stuff(seq, dev);
+	seq_printf(seq, "\n");
 }
 
 /*
@@ -4129,13 +4131,15 @@ static void dev_seq_printf_stats(struct seq_file *seq, struct net_device *dev)
  */
 static int dev_seq_show(struct seq_file *seq, void *v)
 {
-	if (v == SEQ_START_TOKEN)
+	if (v == SEQ_START_TOKEN) {
 		seq_puts(seq, "Inter-|   Receive                            "
 			      "                    |  Transmit\n"
 			      " face |bytes    packets errs drop fifo frame "
 			      "compressed multicast|bytes    packets errs "
-			      "drop fifo colls carrier compressed\n");
-	else
+			      "drop fifo colls carrier compressed");
+		dev_seq_show_qdisc_stuff(seq);
+		seq_puts(seq, "\n");
+	} else
 		dev_seq_printf_stats(seq, v);
 	return 0;
 }
diff --git a/net/sched/sch_generic.c b/net/sched/sch_generic.c
index 67fc573..984a04f 100644
--- a/net/sched/sch_generic.c
+++ b/net/sched/sch_generic.c
@@ -44,7 +44,7 @@ static inline int dev_requeue_skb(struct sk_buff *skb, struct Qdisc *q)
 	skb_dst_force(skb);
 	q->gso_skb = skb;
 	q->qstats.requeues++;
-	q->q.qlen++;	/* it's still part of the queue */
+	qdisc_incr_qlen(q);	/* it's still part of the queue */
 	__netif_schedule(q);
 
 	return 0;
@@ -454,7 +454,7 @@ static int pfifo_fast_enqueue(struct sk_buff *skb, struct Qdisc *qdisc)
 		struct sk_buff_head *list = band2list(priv, band);
 
 		priv->bitmap |= (1 << band);
-		qdisc->q.qlen++;
+		qdisc_incr_qlen(qdisc);
 		return __qdisc_enqueue_tail(skb, qdisc, list);
 	}
 
-- 
1.7.9.7

